pkg/binlog-filter/filter.go

```diff
@@ -0,0 +1,196 @@
+func (b *BinlogEventRule) Valid() error {
+	if len(b.SQLPattern) > 0 {
+		reg, err := regexp.Compile("(?i)" + strings.Join(b.SQLPattern, "|"))
+		if err != nil {
+			return errors.Annotatef(err, "compile regular expression %+v", b.SQLPattern)
+		}
+		b.sqlRegularExp = reg
+	}
+
+	if b.Action != Do && b.Action != Skip {
+		return errors.Errorf("action of binlog event rule %+v should not be empty", b)
+	}
+
+	return nil
+}
+
+// BinlogEvent filters binlog events by given rules
+type BinlogEvent struct {
+	selector.Selector
+}
+
+// NewBinlogEvent returns a binlog event filter
+func NewBinlogEvent(rules []*BinlogEventRule) (*BinlogEvent, error) {
+	b := &BinlogEvent{
+		Selector: selector.NewTrieSelector(),
+	}
+
+	for _, rule := range rules {
+		if err := b.AddRule(rule); err != nil {
+			return nil, errors.Annotatef(err, "initial rule %+v in binlog event filter", rule)
+		}
+	}
+
+	return b, nil
+}
+
+// AddRule adds a rule into binlog event filter
+func (b *BinlogEvent) AddRule(rule *BinlogEventRule) error {
+	err := rule.Valid()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = b.Insert(rule.SchemaPattern, rule.TablePattern, rule, false)
+	if err != nil {
+		return errors.Annotatef(err, "add rule %+v into binlog event filter", rule)
+	}
+
+	return nil
+}
+
+// UpdateRule updates binlog event filter rule
+func (b *BinlogEvent) UpdateRule(rule *BinlogEventRule) error {
+	err := rule.Valid()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = b.Insert(rule.SchemaPattern, rule.TablePattern, rule, true)
+	if err != nil {
+		return errors.Annotatef(err, "update rule %+v into binlog event filter", rule)
+	}
+
+	return nil
+}
+
+// RemoveRule removes a rule from binlog event filter
+func (b *BinlogEvent) RemoveRule(rule *BinlogEventRule) error {
+	err := b.Remove(rule.SchemaPattern, rule.TablePattern)
+	if err != nil {
+		return errors.Annotatef(err, "remove rule %+v", rule)
+	}
+
+	return nil
+}
+
+// Filter filters events or queries by given rules
+// returns action and error
+func (b *BinlogEvent) Filter(schema, table string, dml, ddl EventType, rawQuery string) (ActionType, error) {
+	rules := b.Match(schema, table)
+	if len(rules) == 0 {
+		return Do, nil
+	}
+
+	for _, rule := range rules {
+		binlogEventRule, ok := rule.(*BinlogEventRule)
+		if !ok {
+			return "", errors.NotValidf("rule %+v", rule)
+		}
+
+		matched := false
+		if len(dml) > 0 {
+			matched = b.matchEvent(dml, binlogEventRule.DMLEvent)
+		} else if len(ddl) > 0 {
+			matched = b.matchEvent(ddl, binlogEventRule.DDLEvent)
+		} else if len(rawQuery) > 0 {
+			matched = binlogEventRule.sqlRegularExp.FindStringIndex(rawQuery) != nil
+		} else {
+			if binlogEventRule.Action == Skip { // skip has highest priority
+				return Skip, nil
+			}
+		}
+
+		// skip has highest priority
+		if matched {
+			if binlogEventRule.Action == Skip {
+				return Skip, nil
+			}
+		} else {
+			if binlogEventRule.Action == Do {
+				return Skip, nil
```

Comment:
if I have a `Do` rule, and not set `DMLEvent`, `DDLEvent` and `SQLPattern`, it will `Skip`?

Reply:
Follow the black list definition，`action == skip` -> `matched `;    while `matched` is false, we must use  `action !=  `do`  to keep this logic  formula true

Reply:
that's mean, if I want to have a white list for a whole table, I should not set a rule (`Do` is the default behavior)?

Reply:
I think we should distinguish not-set `DMLEvent` ... from sat but not matched

Reply:
another question. If we have multiple `Do` rules, the first rule mismatched, then it returns `Skip`, but maybe other rules can match it. Should we need to handle this scenario?

Reply:
one problem：what is empty `DMLEvent`.  From the point of setting, the simpler is better, empty is not-setting or disable; but from understanding point, it's the opposite.  Here we must follow strict definitions,  just leave `All` and `None` shortcut

Reply:
I agree "simpler is better".
I just think we should handle the use case
```
BinlogEventRule {
    SchemaPattern: "abc",
    TablePattern: "def",
    DMLEvent: nil,
    DDLEvent: nil,
    SQLPattern: "",
    Action: "Do",
}
```
maybe user expect it will Do the whole table, but the fact is skip it.
to follow strict definitions, maybe we should add some checks (`Valid`) or docs / comment.

Reply:
yep, we must leave a clear explain